upcoming/composable-pipelines-pattern.md

```diff
@@ -0,0 +1,392 @@
+	# Connect our super-minimal little "workflow" consisting
+	# - a base complementer
+	# - a file reader
+	# - our print statements right here in the loop
+	fa_reader = fasta_reader('chr_y.fa')
+	for line in base_complementer(fa_reader):
+		print line
+````
+
+Have a careful look at this line:
+
+````python
+for line in base_complementer(fa_reader):
+	...
+````
+
+We here see how, rather than sending data into `base_complementer`, we instead send a generator object (`fa_reader`), that `base_complementer` will iterate over, as soon as we start iterating over the the latter.
+
+So when executing the main loop in `main()``, it will step by step drive our little pipeline consisting of three parts; a (FASTA) reader, a base complementer generator function and a print-statement at the end.
+
+Again, the lazy evaluation means that one item at a time will be drawn through the whole pipeline for each iteration in the main loop, without any temporary aggregates (lists or dicts) of lines building up between the components, which means the program needs to use very little memory during the process.
+
+## The Generator pattern in Go
+
+Coming into Go, the author was highly intrigued by all the powerful concurrency patterns made possible in this language, elaborated in blog posts such as [the one on "concurrency patterns"](http://blog.golang.org/pipelines) and ["advanced concurrency patterns"](http://blog.golang.org/advanced-go-concurrency-patterns).
+
+Even more interestingly, the simple and straight-forward generator pattern in python was easy to implement in Go too, as shown in [Rob Pike's Go Concurrency pattern slides](https://talks.golang.org/2012/concurrency.slide#25), and also [listed on this site](http://www.golangpatterns.info/concurrency/generators).
+
+The above python code would read something like this in Go, using the generator pattern (leaving out some imports and const definitions, for brevity):
+
+````go
+
+// ...<snip>...
+
+func fileReaderGen(filename string) chan []byte {
+	fileReadChan := make(chan []byte, BUFSIZE)
+	go func() {
+		file, err := os.Open(filename)
+		if err != nil {
+			log.Fatal(err)
+		} else {
+				scan := bufio.NewScanner(file)
+				for scan.Scan() {
+					// Write to the channel we will return, while copying
+					// the buffer slice (which is re-used) ...
+					fileReadChan <- append([]byte(nil), scan.Bytes()...)
+		}
+		close(fileReadChan)
+		fmt.Println("Closed file reader channel")
+		}
+	}()
+	return fileReadChan
+}
+
+// Translation table, used in the base complementer
+// function below
+var baseConv = [256]byte{
+	'A': 'T',
+	'T': 'A',
+	'C': 'G',
+	'G': 'C',
+	'N': 'N',
+	'\n' : '\n',
+}
+
+func baseComplementGen(inChan chan []byte) chan []byte {
+	returnChan := make(chan []byte, BUFSIZE)
+	go func() {
+		for sequence := range inChan {
+			if sequence[0] != '>' {
+				for pos := range sequence {
+					sequence[pos] = baseConv[sequence[pos]]
+				}
+			}
+			returnChan <- append([]byte(nil), sequence...)
```

Comment:
yah, i still don't understand what this is trying to do. why not just `<- sequence`?